### DISCUSSION TITLE
Uvicorn fails "quietly" upon a circular import error

### DISCUSSION THREAD

**Discussion by mikeedjones** [Kludex/uvicorn#2034] [Potential Issue] [answered]
From this [comment](https://github.com/tiangolo/fastapi/discussions/9838#discussioncomment-6391673) and discussion.

When there is a circular import in the app, uvicorn reports `ERROR:    Error loading ASGI app. Could not import module "main".` without exposing the stacktrace which would be reported if you ran the code using `python main.py`. MRE below. 

```python
# main.py
from bar import Bar

class Foo:
    pass

async def app(scope, receive, send):
    assert scope['type'] == 'http'

```

```python
#bar.py
from main import Foo

class Bar:
    pass

```

Running with 
```
uvicorn main:app
``` 
only results in 
```
ERROR:    Error loading ASGI app. Could not import module "main".
```
whereas running 
```
python main.py
``` 
reports the far more useful 
```
ImportError: cannot import name 'Bar' from partially initialized module 'bar' (most likely due to a circular import) (/workspaces/fastapi/bar.py)
```

Not sure how you'd start debugging this one I'm afraid - or if its even a "bug". For comparison syntax errors are reported correctly.

**Comment by Kludex** [comment, marked as answer]
Thanks.

https://github.com/encode/uvicorn/issues/2035

**Comment by Markundi** [comment]
Hello, 

Could you please suggest me on this ( Kafka FAST API)

##########main.py ################################
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}

---------------------------------------------------------

error :
C:\Users\Satish\microservice-kafka-project\ms_01>docker compose logs api -f
ms_01  | Skipping virtualenv creation, as specified in config file.
ms_01  | INFO:     Will watch for changes in these directories: ['/code']
ms_01  | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
ms_01  | INFO:     Started reloader process [1] using StatReload
ms_01  | ERROR:    **Error loading ASGI app. Could not import module "ms_01.main".**

![image](https://github.com/user-attachments/assets/c2d18471-341c-47b0-ace7-90ace4a2ff74)
